### ISSUE TITLE
error  in gsw_rr68_interp_sa_ct

### ISSUE THREAD

**Issue by dankelley** [TEOS-10/GSW-C#17]
Presently, line 7813 of `gsw_oceanographic_toolbox.c` (link: https://github.com/TEOS-10/GSW-C/blob/master/gsw_oceanographic_toolbox.c#L7813)  reads
```
	        if (p_i[i] == p[j]) {
```
but comparing floating-point numbers is problematic, owing to binary representation, and I think this would be better as (e.g. judging sameness as being within 1mm in depth)
```
               if (fabs(p_i[i] - p[j]) < 0.001) {
```
To whoever is maintaining the C (and possibly fortran and matlab versions):

I ran into this because I was getting NaN values in my GSW-R code, which uses a copy of `gsw_oceanographic_toolbox.c`.  Making the change removes the NaN problems.

 I had a very quick look at the fortran code, but things are different enough there that I could not see whether the error was repeated there, also.

I did a little grepping for similar things and, ignoring cases that are at a glance about integer comparisons, 
```
git grep -n " == " src/gsw_oceanographic_toolbox.c
```
gives as follows. My head is not inside the code, so I don't know if these are problems also.
```
src/gsw_oceanographic_toolbox.c:3717:	    if (p[ibottle] == p_ref) {
src/gsw_oceanographic_toolbox.c:3722:	if ((dp_max <= max_dp_i) && (p[0] == 0.0) && (ipref >= 0)) {
src/gsw_oceanographic_toolbox.c:3841:	            if (p[ibottle] == p_ref) ibpr = p_cnt-1;
src/gsw_oceanographic_toolbox.c:3865:	        if (p[nz-1] == p_ref) ibpr = p_cnt-1;
src/gsw_oceanographic_toolbox.c:10395:	    if (z == x[k+1])
```

PS. my concern about whether this problem is reproduced in some of the other comparisons, and my hope that someone might look also at the fortran and matlab code for similar problems, explains why I'm providing these notes in an issue report, as opposed to making a fork and then a pull request.

Best regards, Dan.

**Comment by hylandg** [comment]
Hi Dan,

That section of code looks fairly innocuous ...
```C
	        if (p_i[i] == p[j]) {
	            sa_i[i] = sa[j];
	            ct_i[i] = ct[j];
                }
```
I suspect the `NaN` is being generated earlier in the function (perhaps even in the `rr68_interp_section` function) and is then being overwritten by the mods that you made.

Can you supply me with the smallest set of `sa, ct, p` and `pt_i` values that reproduces the error? Then I can trace through the code to see where the problem lies. Thanks.

**Comment by dankelley** [comment]
Thanks, Glenn.

I put an example at https://www.dropbox.com/sh/rvgb7s36i8l5mao/AAB0yBLJHko-YGYYL8QAhY1ya?dl=0 and to run it, type
```
make
```
on a unix-type machine.

The test is in `issue1.c`. The other `.c` files and `.h` files are copied from `GSW-C`, commit 13ec1cb2e60e472a319d74b78013aa01990d24e9; below is a check that I'm using those `.c` and `.h` files.
```
$ pushd ~/git/GSW-C ; git log -1
~/git/GSW-C ~/Dropbox/gsw_nan
commit 13ec1cb2e60e472a319d74b78013aa01990d24e9 (HEAD -> master, origin/master, origin/HEAD)
Merge: 6378b71 6ffdd1a
Author: Glenn Hyland <Glenn.Hyland@aad.gov.au>
Date:   Thu Apr 20 08:38:13 2017 +1000

    Merge pull request #16 from efiring/add_gitignore
    
    Add .gitignore


$ for f in *.c *.h; do md5 $f; md5 ~/git/GSW-C/$f ; done
MD5 (gsw_oceanographic_toolbox.c) = a4bfa0d23565c0990a9e5e75ca64aa4f
MD5 (/Users/kelley/git/GSW-C/gsw_oceanographic_toolbox.c) = a4bfa0d23565c0990a9e5e75ca64aa4f
MD5 (gsw_saar.c) = 7329352641859be2b029a1428e624e51
MD5 (/Users/kelley/git/GSW-C/gsw_saar.c) = 7329352641859be2b029a1428e624e51
MD5 (gsw_saar_data.c) = c1ba4e4d9a2c67209d6cb78342b50c81
MD5 (/Users/kelley/git/GSW-C/gsw_saar_data.c) = c1ba4e4d9a2c67209d6cb78342b50c81
MD5 (issue1.c) = 43a80cf7e84fe0a3ad116b38bf5ec4dc
md5: /Users/kelley/git/GSW-C/issue1.c: No such file or directory
MD5 (gsw_internal_const.h) = 51752d8931de194a12e7f66464b40848
MD5 (/Users/kelley/git/GSW-C/gsw_internal_const.h) = 51752d8931de194a12e7f66464b40848
MD5 (gswteos-10.h) = 3c2d8105e942f3aeeb19a44130c28d10
MD5 (/Users/kelley/git/GSW-C/gswteos-10.h) = 3c2d8105e942f3aeeb19a44130c28d10
```

**Comment by hylandg** [comment]
Thanks for the example code Dan. The code allowed me to track down a couple of problems and provide fixes to both the C (13ab33c) and Fortran codes (https://github.com/TEOS-10/GSW-Fortran/commit/c6bec537604b5f84dce053705aeeb4aee84308fd). Your example now produces sensible numbers.

There was a problem in two locations (plus an additional fault with the Fortran code). Firstly, in a helper routine `p_sequence` called by `rr68_interp_section`. This function generates a sequence of pressures from `p1` (exclusive) to `p2` (inclusive), but did it in a way that meant the value of `p2 ` could be included with roundoff. So by changing the code from
```C
       for (i=1; i<=n; i++)
           pseq[i-1] = p1+pstep*i;
```
to
```C
       for (i=0; i<n; i++)
           pseq[i] = p2-pstep*(n-1-i);
```
then the value of `p2` is included exactly. This fix alone solves Dan's problem, because now the equivalence test on line 7813 will always work, and the `Nans` generated elsewhere in the code will be replaced with sensible values.

Secondly, the `Nans` were being generated by a logic flaw in `gsw_rr68_interp_sa_ct` that stems back to the original Matlab code. Fortunately the Matlab code has moved on to v3.06, which replaces `gsw_rr68_interp_sa_ct` with another interpolation scheme, and so is not affected by this problem anymore (but earlier versions still are). The problem is caused by
```C
        floor(x) == ceil(x)
```
when `x` is a whole number. So the calculation of the indices `ip_1, ip_2, ip_3, ip_4` in function `rr68_interp_section` needed some extra code to check for this condition (see the commits for the simple code inclusions).

And, hey presto - no more `Nans` (caused by 'division by zero'). In fact, with this fix in place, the equivalence test on line 7813 becomes redundant, and the code could be probably be removed.
